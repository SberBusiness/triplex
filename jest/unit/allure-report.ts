import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
