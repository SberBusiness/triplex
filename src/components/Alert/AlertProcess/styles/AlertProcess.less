@import 'src/styles/style';

@alert-list-marker-width: @metric-12;
@border-width: 1px;

.alertProcess {
    display: flex;
    padding: @metric-16 - @border-width;
    border-width: @border-width;
    border-style: solid;
    border-radius: @radius-8;
    .b1-text-main(@color-graphite-01);
    color: var(--triplex-AlertProcess-Color);

    .themeIcon {
        padding-top: 2px;
        box-sizing: border-box;
        height: 20px;
        width: 20px;
    }

    &.alertTypeError {
        background: var(--triplex-AlertProcess-Error_Background);
        border-color: var(--triplex-AlertProcess-Error_Border_Color);
    }

    &.alertTypeInfo {
        background: var(--triplex-AlertProcess-Info_Background);
        border-color: var(--triplex-AlertProcess-Info_Border_Color);
    }

    &.alertTypeWarning {
        background: var(--triplex-AlertProcess-Warning_Background);
        border-color: var(--triplex-AlertProcess-Warning_Border_Color);
    }

    &.alertTypeFeature {
        background: var(--triplex-AlertProcess-Feature_Background);
        border-color: var(--triplex-AlertProcess-Feature_Border_Color);

        .themeIcon {
            padding-top: 0;
        }
    }

    .alertProcessContentBlock {
        flex: 1;
        margin-left: @metric-8;

        .alertProcessContentHeader {
            margin-bottom: @metric-8;
            .font-medium();
            .font-size-inline(@font-size-14);

            text-transform: none;
            color: var(--triplex-AlertProcess-Color);
        }
    }

    .closeButton {
        margin-left: @metric-8;
        height: @metric-16;
    }
}
