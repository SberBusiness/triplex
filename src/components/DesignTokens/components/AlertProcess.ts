import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента AlertProcess.
export const designTokensComponentsAlertProcessKeys = [
    'Color',
    'Error_Background',
    'Error_Border_Color',
    'Feature_Background',
    'Feature_Border_Color',
    'Info_Background',
    'Info_Border_Color',
    'Warning_Background',
    'Warning_Border_Color',
] as const;
// Тип, содержащий названия токенов компонента AlertProcess.
export type TDesignTokensComponentsAlertProcessKeys = (typeof designTokensComponentsAlertProcessKeys)[number];
// Тип, содержащий названия токенов компонента AlertProcess и их значения.
export type TDesignTokensComponentsAlertProcessValue = Record<TDesignTokensComponentsAlertProcessKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента AlertProcess и их значения в светлой и темной теме.
export type TDesignTokensComponentsAlertProcessValues = Record<TDesignTokensComponentsAlertProcessKeys, TDesignTokenValues>;
// Тип локальных токенов компонента AlertProcess.
export type TDesignTokensComponentsAlertProcess = {AlertProcess: TDesignTokensComponentsAlertProcessValue};

// Токены компонента AlertProcess в светлой и темной темах.
export const AlertProcess_Tokens: TDesignTokensComponentsAlertProcessValues = {
    Color: [{ref: 'Basic.700'}, {ref: 'Neutral.100'}], // var(--triplex-AlertProcess-Color)
    Error_Background: [{ref: 'Error.100'}, {ref: 'Basic.0'}], // var(--triplex-AlertProcess-Error_Background)
    Error_Border_Color: [{ref: 'Error.700'}, {ref: 'Error.500'}], // var(--triplex-AlertProcess-Error_Border_Color)
    Feature_Background: [{ref: 'Primary.100'}, {ref: 'Basic.0'}], // var(--triplex-AlertProcess-Feature_Background)
    Feature_Border_Color: [{ref: 'Primary.500'}, {ref: 'Primary.500'}], // var(--triplex-AlertProcess-Feature_Border_Color)
    Info_Background: [{ref: 'Info.100'}, {ref: 'Basic.0'}], // var(--triplex-AlertProcess-Info_Background)
    Info_Border_Color: [{ref: 'Info.700'}, {ref: 'Info.500'}], // var(--triplex-AlertProcess-Info_Border_Color)
    Warning_Background: [{ref: 'Warning.100'}, {ref: 'Basic.0'}], // var(--triplex-AlertProcess-Warning_Background)
    Warning_Border_Color: [{ref: 'Warning.700'}, {ref: 'Warning.700'}], // var(--triplex-AlertProcess-Warning_Border_Color)
};
