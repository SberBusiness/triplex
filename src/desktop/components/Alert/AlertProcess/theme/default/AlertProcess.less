@import 'src/desktop/common/theme/default.less';

@alert-list-marker-width: @metric-12;
@border-width: 1px;

.alertProcess {
    display: flex;
    padding: @metric-16 - @border-width;
    border-width: @border-width;
    border-style: solid;
    border-radius: @radius-8;
    .b1-text-main(@color-graphite-01);

    .themeIcon {
        padding-top: 2px;
        box-sizing: border-box;
        height: 20px;
        width: 20px;
    }

    &.alertTypeError {
        background: @color-ruby-05;
        border-color: @color-ruby-02;
    }

    &.alertTypeInfo {
        background: @color-sky-05;
        border-color: @color-sky-02;
    }

    &.alertTypeWarning {
        background: @color-pumpkin-05;
        border-color: @color-pumpkin-02;
    }

    &.alertTypeFeature {
        background: @color-malachite-05;
        border-color: @color-malachite-03;

        .themeIcon {
            padding-top: 0;
        }
    }

    .alertProcessContentBlock {
        flex: 1;
        margin-left: @metric-8;

        .alertProcessContentHeader {
            margin-bottom: @metric-8;
            .b3-text-controls-medium(@color-graphite-01);
        }

        ul {
            list-style: none;
            margin: 0 0 0 @alert-list-marker-width;
            padding: 0;

            li {
                color: @color-graphite-01;

                &::before {
                    content: '\2022';
                    display: inline-block;
                    width: @alert-list-marker-width;
                    margin-left: -@alert-list-marker-width;
                    font-weight: bold;
                }

                &:not(:last-child) {
                    margin-bottom: @metric-8;
                }
            }
        }
    }

    .closeButton {
        margin-left: @metric-8;
        height: @metric-16;
    }
}
