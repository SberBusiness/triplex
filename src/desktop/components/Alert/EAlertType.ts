/** Тип компонентов-предупреждений. */
export enum EAlertType {
    INFO = 'info',
    WARNING = 'warning',
    ERROR = 'error',
    SYSTEM = 'system',
    FEATURE = 'feature',
}
