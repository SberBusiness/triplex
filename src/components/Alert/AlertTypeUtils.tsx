import React from 'react';
import {EAlertType} from '@sberbusiness/triplex/components/Alert/EAlertType';
import {InfoStsIcon16} from '@sberbusiness/icons/InfoStsIcon16';
import {WarningStsIcon16} from '@sberbusiness/icons/WarningStsIcon16';
import {ErrorStsIcon16} from '@sberbusiness/icons/ErrorStsIcon16';
import {SystemStsIcon16} from '@sberbusiness/icons/SystemStsIcon16';
import {ReportstogovernmentPrdIcon20} from '@sberbusiness/icons/ReportstogovernmentPrdIcon20';

/** Получить иконку по типу предупреждения. */
export function renderDefaultIcon(type: EAlertType): JSX.Element {
    switch (type) {
        case EAlertType.INFO:
            return <InfoStsIcon16 />;
        case EAlertType.WARNING:
            return <WarningStsIcon16 />;
        case EAlertType.ERROR:
            return <ErrorStsIcon16 />;
        case EAlertType.SYSTEM:
            return <SystemStsIcon16 />;
        case EAlertType.FEATURE:
            return (
                <div className="hoverable active">
                    <ReportstogovernmentPrdIcon20 />
                </div>
            );
    }
}

/** Получить класс по типу предупреждения. */
export const alertTypeToClassNameMap = {
    [EAlertType.INFO]: 'cssClass[alertTypeInfo]',
    [EAlertType.WARNING]: 'cssClass[alertTypeWarning]',
    [EAlertType.ERROR]: 'cssClass[alertTypeError]',
    [EAlertType.SYSTEM]: 'cssClass[alertTypeSystem]',
    [EAlertType.FEATURE]: 'cssClass[alertTypeFeature]',
};
