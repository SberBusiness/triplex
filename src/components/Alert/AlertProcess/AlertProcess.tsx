import React, {useState} from 'react';
import {EAlertType} from '@sberbusiness/triplex/components/Alert/EAlertType';
import {AlertProcessHeader} from '@sberbusiness/triplex/components/Alert/AlertProcess/components/AlertProcessHeader';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {alertTypeToClassNameMap, renderDefaultIcon} from '@sberbusiness/triplex/components/Alert/AlertTypeUtils';
import {ButtonIcon} from '@sberbusiness/triplex/components/Button/ButtonIcon';
import {ClosenotificationSrvxIcon16} from '@sberbusiness/icons/ClosenotificationSrvxIcon16';

/** Свойства компонента AlertProcess. */
export interface IAlertProcessProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Тип предупреждения. */
    type: EAlertType;
    /** Модификатор возможности закрытия предупреждения. */
    closable?: boolean;
    /** Функция обработки закрытия. */
    onClose?: () => void;
    /** Рендер-функция заголовка. */
    renderHeader?: () => JSX.Element;
    /** Рендер-функция иконки. */
    renderIcon?: () => JSX.Element;
}

/** Компонент процессного предупреждения. */
export const AlertProcess = Object.assign(
    React.forwardRef<HTMLDivElement, IAlertProcessProps>(function AlertProcess(
        {children, className, type, closable, onClose, renderHeader, renderIcon, ...rest},
        ref
    ) {
        const [closed, setClosed] = useState(false);

        if (closed) {
            return null;
        }

        const handleClose = () => {
            setClosed(true);

            onClose?.();
        };

        return (
            <div
                className={classnames('cssClass[alertProcess]', alertTypeToClassNameMap[type], className)}
                {...rest}
                data-tx={process.env.npm_package_version}
                ref={ref}
            >
                <div className="cssClass[themeIcon]">{renderIcon ? renderIcon() : renderDefaultIcon(type)}</div>
                <div className="cssClass[alertProcessContentBlock]">
                    {renderHeader?.()}
                    {children}
                </div>
                {closable && (
                    <div className="cssClass[closeButton]">
                        <ButtonIcon onClick={handleClose}>
                            <ClosenotificationSrvxIcon16 />
                        </ButtonIcon>
                    </div>
                )}
            </div>
        );
    }),
    {
        Header: AlertProcessHeader,
    }
);

AlertProcess.displayName = 'AlertProcess';
AlertProcess.Header = AlertProcessHeader;
