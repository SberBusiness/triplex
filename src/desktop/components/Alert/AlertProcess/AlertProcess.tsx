import React, {useState, useEffect} from 'react';
import {EAlertType} from '@sberbusiness/triplex/desktop/components/Alert/EAlertType';
import {AlertProcessHeader} from '@sberbusiness/triplex/desktop/components/Alert/AlertProcess/components/AlertProcessHeader';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import {mapAlertTypeToClassName, renderDefaultIcon} from '@sberbusiness/triplex/desktop/components/Alert/AlertTypeUtils';
import {ButtonIcon} from '@sberbusiness/triplex/desktop/components/Button/ButtonIcon';
import {ClosenotificationSrvxIcon16} from '@sberbusiness/icons/ClosenotificationSrvxIcon16';

/** @deprecated Устаревший интерфейс компонента AlertProcess. */
export interface IDeprecatedAlertProcessProps extends Omit<IAlertProcessProps, 'type'> {
    /** Текст внутри компонента. */
    text: React.ReactNode;
}

/** Интерфейс общих свойств компонента AlertProcess. */
export interface IAlertProcessProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Тип предупреждения. */
    type: Exclude<EAlertType, EAlertType.SYSTEM>;
    /** Модификатор возможности закрытия предупреждения. */
    closable?: boolean;
    /** Функция обработки закрытия. */
    onClose?: () => void;
    /** Рендер-функция заголовка. */
    renderHeader?: () => JSX.Element;
    /** Рендер-функция иконки. */
    renderIcon?: () => JSX.Element;
}

/** Внутренние составляющие процессного предупреждения. */
interface IAlertProcessComposition {
    Header: typeof AlertProcessHeader;
}

/** Компонент процессного предупреждения. */
export const AlertProcess: React.FC<IAlertProcessProps> & IAlertProcessComposition = ({
    children,
    className,
    type,
    closable,
    onClose,
    renderHeader,
    renderIcon,
    ...rest
}) => {
    const [isOpen, setIsOpen] = useState(true);

    useEffect(() => {
        if (!isOpen) {
            onClose?.();
        }
    }, [isOpen]);

    const handleClose = () => setIsOpen(false);

    const render = () => (
        <div className={classnames('cssClass[alertProcess]', mapAlertTypeToClassName(type), className)} {...rest}>
            <div className="cssClass[themeIcon]">{renderIcon ? renderIcon() : renderDefaultIcon(type)}</div>
            <div className="cssClass[alertProcessContentBlock]">
                {renderHeader?.()}
                {children}
            </div>
            {closable && (
                <div className="cssClass[closeButton]">
                    <ButtonIcon onClick={handleClose}>
                        <ClosenotificationSrvxIcon16 />
                    </ButtonIcon>
                </div>
            )}
        </div>
    );

    return isOpen ? render() : null;
};

AlertProcess.displayName = 'AlertProcess';
AlertProcess.Header = AlertProcessHeader;
