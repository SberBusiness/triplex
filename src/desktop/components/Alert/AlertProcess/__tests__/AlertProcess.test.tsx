import React from 'react';
import renderer from 'react-test-renderer';
import {allure} from '@jest/unit/allure-report';
import {AlertProcess} from '@sberbusiness/triplex/desktop/components/Alert/AlertProcess/AlertProcess';
import {EAlertType} from '@sberbusiness/triplex/desktop/components/Alert/EAlertType';

jest.mock('@sberbusiness/icons/InfoStsIcon16', () => ({
    InfoStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/WarningStsIcon16', () => ({
    WarningStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/ErrorStsIcon16', () => ({
    ErrorStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/SystemStsIcon16', () => ({
    SystemStsIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/ReportstogovernmentPrdIcon20', () => ({
    ReportstogovernmentPrdIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/ClosenotificationSrvxIcon16', () => ({
    ClosenotificationSrvxIcon16: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

describe('AlertProcess', () => {
    beforeEach(() => {
        allure.feature('AlertProcess');
    });

    const text = 'Sample Text';

    it('type info renders correctly', () => {
        const tree = renderer.create(<AlertProcess type={EAlertType.INFO}>{text}</AlertProcess>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('type warning renders correctly', () => {
        const tree = renderer.create(<AlertProcess type={EAlertType.WARNING}>{text}</AlertProcess>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('type error renders correctly', () => {
        const tree = renderer.create(<AlertProcess type={EAlertType.ERROR}>{text}</AlertProcess>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('type system renders correctly', () => {
        const tree = renderer.create(<AlertProcess type={EAlertType.FEATURE}>{text}</AlertProcess>).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
