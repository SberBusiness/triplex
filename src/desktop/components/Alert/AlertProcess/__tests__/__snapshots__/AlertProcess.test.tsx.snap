// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertProcess type error renders correctly 1`] = `
<div
  className="cssClass[alertProcess] cssClass[alertTypeError]"
>
  <div
    className="cssClass[themeIcon]"
  >
    <svg />
  </div>
  <div
    className="cssClass[alertProcessContentBlock]"
  >
    Sample Text
  </div>
</div>
`;

exports[`AlertProcess type info renders correctly 1`] = `
<div
  className="cssClass[alertProcess] cssClass[alertTypeInfo]"
>
  <div
    className="cssClass[themeIcon]"
  >
    <svg />
  </div>
  <div
    className="cssClass[alertProcessContentBlock]"
  >
    Sample Text
  </div>
</div>
`;

exports[`AlertProcess type system renders correctly 1`] = `
<div
  className="cssClass[alertProcess] cssClass[alertTypeFeature]"
>
  <div
    className="cssClass[themeIcon]"
  >
    <div
      className="hoverable active"
    >
      <svg />
    </div>
  </div>
  <div
    className="cssClass[alertProcessContentBlock]"
  >
    Sample Text
  </div>
</div>
`;

exports[`AlertProcess type warning renders correctly 1`] = `
<div
  className="cssClass[alertProcess] cssClass[alertTypeWarning]"
>
  <div
    className="cssClass[themeIcon]"
  >
    <svg />
  </div>
  <div
    className="cssClass[alertProcessContentBlock]"
  >
    Sample Text
  </div>
</div>
`;
